### PR TITLE
ZTP: Add tpm disk encryption with PCRs and solution for upgrade

### PR DESCRIPTION
**Pull request by edcdavid**
Do not Merge! This is work in progress

This PR add the following:
1. a new MC with 2 systemd units
1.a. that runs just before shutdown and can detect whether changes that impact PCRs value is going to occur in the next reboot. if so, it will add a temporary LUKS slot with alternative value (by default: tpm only without PCRs)
1.b. that runs after startup, and can detect whether we need to remove the temporary slot and rebind with new PCRs value
2. siteconfig support for TPM with PCRs disk encryption

### CODE REVIEW

File: ztp/source-crs/extra-manifest/disk-encryption.yaml.tmpl [hunk 1]
```diff
@@ -14,18 +14,38 @@ spec:
...
             tang:
             {{- range .Tang }}
                 - url: "{{ .URL }}"
                   thumbprint: "{{ .Thumbprint }}"
             {{- end }}
+          {{- else if eq .Type "tpm2+nbde" }}
+            custom:
+              pin: "sss"
+              needsNetwork: true
+              config: '{"t":2,"pins":{"tpm2":[{}],"sss":[{"t":1,"pins":{"tang":[
```
Comment (lack) [inline ztp/source-crs/extra-manifest/disk-encryption.yaml.tmpl:33]: This should probably also add PRS at if .Tmp2.PCRList is set, right?

**Comment by yuvalk** [comment]
@edcdavid please fix commits so that each commit is atomic and complete (a single complete change)
make sure they all have the appropriate prefix of `ztp:` https://github.com/openshift-kni/cnf-features-deploy/blob/master/.githooks/components.sh#L8
then make sure your commit msg follow best practices. mainly, first line is "subject" followed by an empty line. then have the first paragraph (as well as the subject line) explain the WHY of the change (not the WHAT which can be observed from the content)
I tend to add additional paragraphs if needed with more info on the change, explaining decisions, providing references, etc.

try to make it like a math proof where you first lay the foundations then build on top of them. 
In that regards, I'd strive to have the new MC in it's own commit. certainly separated from the siteconfig changes

File: ztp/ran-crd/site-config-crd.yaml [hunk 2]
```diff
@@ -792,8 +801,7 @@ spec:
                                   type: string
                                 thumbprint:
                                   description: Tang server trusted thumbprint
-                                  type: string
-                      crTemplates:
+                                  type: string                      crTemplates:
```
Comment (lack) [inline ztp/ran-crd/site-config-crd.yaml:804]: This looks like an error, maybe?

Comment (edcdavid) [reply ztp/ran-crd/site-config-crd.yaml]: good catch! fixed

**Review by yuvalk — COMMENTED** [review]
can you make commit subject (first line) short enough so it'll fit in the line?

**Review by yuvalk — COMMENTED** [review]
can you split this commit so the ztp tpm2 part is separated from the pcrs stuff?

File: ztp/ran-crd/site-config-cr-ex.yaml [hunk 3]
```diff
@@ -34,7 +34,7 @@ spec:
       - NTP.server1
       - 10.16.231.22
     diskEncryption:
-      type: "nbde"
+      type: "tpm2+nbde"
```
Comment (yuvalk) [inline ztp/ran-crd/site-config-cr-ex.yaml:37]: I'm not sure if we wanna change the example
@lack ?

File: ztp/siteconfig-generator-kustomize-plugin/testSiteConfig/site1-sno-du-tpm2-nbde.yaml [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    machineNetwork:
+      - cidr: 10.16.231.0/24
+    serviceNetwork:
+      - 172.30.0.0/16
+    manifestsConfig:
+      ntpServer: "pool.ntp.org"
+    additionalNTPSources:
+      - NTP.server1
+      - 10.16.231.22
+    ignitionConfigOverride: "igen"
```
Comment (yuvalk) [inline ztp/siteconfig-generator-kustomize-plugin/testSiteConfig/site1-sno-du-tpm2-nbde.yaml:37]: I think this file should contain ONLY the disk encryption stuff
kustomize will bring the rest from the other files...

File: ztp/siteconfig-generator-kustomize-plugin/testSiteConfig/site2-sno-du.yaml [hunk 5]
```diff
@@ -31,6 +31,10 @@ spec:
       - NTP.server1
       - 10.16.231.22
     ignitionConfigOverride: "igen"
+    diskEncryption:
+      type: "tpm2"
+      tpm2:
+        pcrList: "1,8"
```
Comment (yuvalk) [inline ztp/siteconfig-generator-kustomize-plugin/testSiteConfig/site2-sno-du.yaml:37]: I think we need to add more sites
so we can test all 4 options (none, tpm2, nbde, tpm2+nbde)

File: ztp/source-crs/extra-manifest/disk-encryption.yaml.tmpl [hunk 6]
```diff
@@ -34,13 +54,15 @@ spec:
             [Unit]
             Description=Sync the main system network configuration into initramfs for consistency at boot
             After=NetworkManager.service
+            ConditionKernelCommandLine=ignition.firstboot
 
             [Service]
             Type=simple
-            ExecStart=rpm-ostree ex initramfs-etc --track /etc/NetworkManager
+            ExecStart=/bin/bash -c "/bin/rpm-ostree initramfs-etc --track /etc/NetworkManager && /bin/rpm-ostree status && /bin/systemctl reboot"
```
Comment (yuvalk) [inline ztp/source-crs/extra-manifest/disk-encryption.yaml.tmpl:61]: can you please move this to a separate commit that would be about fixing NBDE?

File: ztp/extra-manifests-builder/01-disk-encryption-pcr-rebind/luks-helpers.sh [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        IFS="|" read -ra values <<<"$line"
+        reservedSlotPresent=${values[0]}
+        device=${values[1]}
+        slotNumber=${values[2]}
+        pcrIDs=${values[3]}
+        clevisConfig=${values[4]}
+        logInfo "reservedSlot=$reservedSlotPresent device=$device slot=$slotNumber with PCR IDs=$pcrIDs and clevis config=$clevisConfig"
+        if [ -n "$pcrIDs" ]; then
+            logDebug "before applying command: $(/usr/bin/tpm2_pcrread sha256:$pcrIDs)"
+            $1 $reservedSlotPresent $device $slotNumber $pcrIDs $clevisConfig || true
```
Comment (yuvalk) [inline ztp/extra-manifests-builder/01-disk-encryption-pcr-rebind/luks-helpers.sh:126]: I'd rather you'll add that as a variable at the top of the function and then use the generic function name

File: ztp/extra-manifests-builder/01-disk-encryption-pcr-rebind/luks-helpers.sh [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+addReservedSlot() {
+    logInfo "reservedSlotPresent=$1 device=$2 slot=$3 with PCR IDs=$4 and clevis config=$5"
+    if [ $1 == $TRUE ]; then
+        logInfo "reserve slot already present, no need to add again"
+        clevis luks list -d $2 || true
+        return
+    fi
+    logInfo "adding reserved slot on device=$device"
+    ANYPASS=$(openssl rand -base64 21)
+    echo -e "$ANYPASS\n" | clevis luks bind -s $RESERVED_SLOT -d $2 tpm2 '{}' || true
```
Comment (yuvalk) [inline ztp/extra-manifests-builder/01-disk-encryption-pcr-rebind/luks-helpers.sh:60]: I think the reserved slot content should be configurable (and I dont mind tpm2 only being the default)
but for tpm2+nbde we should at least be able to explain customers how to change the reserved slot config so it'll retain the nbde information.


an even nicer way can be if you copy that from the original slot but that can be tricky.


the other option, is maybe ditch tpm2+nbde
might make it all cleaner
then we can revisit that and how to properly do it in the future
@lack WDYT?

Comment (yuvalk) [reply ztp/extra-manifests-builder/01-disk-encryption-pcr-rebind/luks-helpers.sh:60]: the script already engage the reserved slot only if there are PCRs.. (so there is tpm2..)

File: ztp/extra-manifests-builder/01-disk-encryption-pcr-rebind/luks-helpers.sh [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    clevis luks list -d $2 || true
+}
+
+# remove the temporary key in the reserved slot to enable PCR protection
+removeReservedSlot() {
+    logInfo "removing luks reserved slot 31 in disk $1"
+    echo "clevis luks unbind -s $RESERVED_SLOT -d $1 -f" | sed 's@/@\/@g' | bash || true
```
Comment (yuvalk) [inline ztp/extra-manifests-builder/01-disk-encryption-pcr-rebind/luks-helpers.sh:67]: can you rewrite it without the sed?

Comment (yuvalk) [reply ztp/extra-manifests-builder/01-disk-encryption-pcr-rebind/luks-helpers.sh:67]: maybe 
`-d $1` -> `-d \"$1\"`

Comment (yuvalk) [reply ztp/extra-manifests-builder/01-disk-encryption-pcr-rebind/luks-helpers.sh:67]: if not, maybe worth adding a comment it's because of clevis oddity

**Comment by yuvalk** [comment]
I suggest adding more test coverage

File: ztp/source-crs/extra-manifest/disk-encryption.yaml.tmpl (hunk 1 above)
Comment (edcdavid) [reply ztp/source-crs/extra-manifest/disk-encryption.yaml.tmpl]: removed support for nbde+tpm2 for this first PR

File: ztp/ran-crd/site-config-cr-ex.yaml (hunk 3 above)
Comment (edcdavid) [reply ztp/ran-crd/site-config-cr-ex.yaml]: fixed

File: ztp/siteconfig-generator-kustomize-plugin/testSiteConfig/site1-sno-du-tpm2-nbde.yaml (hunk 4 above)
Comment (edcdavid) [reply ztp/siteconfig-generator-kustomize-plugin/testSiteConfig/site1-sno-du-tpm2-nbde.yaml]: fixed

File: ztp/siteconfig-generator-kustomize-plugin/testSiteConfig/site2-sno-du.yaml (hunk 5 above)
Comment (edcdavid) [reply ztp/siteconfig-generator-kustomize-plugin/testSiteConfig/site2-sno-du.yaml]: fixed

File: ztp/source-crs/extra-manifest/disk-encryption.yaml.tmpl (hunk 6 above)
Comment (edcdavid) [reply ztp/source-crs/extra-manifest/disk-encryption.yaml.tmpl:50]: fixed

File: ztp/extra-manifests-builder/01-disk-encryption-pcr-rebind/luks-helpers.sh (hunk 7 above)
Comment (edcdavid) [reply ztp/extra-manifests-builder/01-disk-encryption-pcr-rebind/luks-helpers.sh]: fixed

File: ztp/extra-manifests-builder/01-disk-encryption-pcr-rebind/luks-helpers.sh (hunk 8 above)
Comment (edcdavid) [reply ztp/extra-manifests-builder/01-disk-encryption-pcr-rebind/luks-helpers.sh]: removed the nbde+tpm2 scenario in this PR

File: ztp/extra-manifests-builder/01-disk-encryption-pcr-rebind/luks-helpers.sh (hunk 9 above)
Comment (edcdavid) [reply ztp/extra-manifests-builder/01-disk-encryption-pcr-rebind/luks-helpers.sh]: fixed

File: ztp/siteconfig-generator-kustomize-plugin/testSiteConfig/site1-sno-du-tpm2-nbde.yaml [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    diskEncryption:
+      type: "tpm2+nbde"
+      tpm2:
+        pcrList: 1,8
```
[the remark this answers is not in the document]
Comment (edcdavid) [reply ztp/siteconfig-generator-kustomize-plugin/testSiteConfig/site1-sno-du-tpm2-nbde.yaml]: fixed

**Comment by yuvalk** [comment]
for someone who doesnt know the materia, it might be very hard to understand.
I suggest adding a README.md file with explanation of the flow

**Comment by yuvalk** [comment]
@lack are we good with approving this?

**Comment by yuvalk** [comment]
@edcdavid can you please fix the commit history
remove the merge commit
squash the bug fix into the right commit, so it wont be introduced in the first place

**Comment by edcdavid** [comment]
Fixed merge issue and regression. Going through a full re-test.

**Review by lack — APPROVED** [review]
